limits/filetransfer/limits.go

```diff
@@ -0,0 +1,391 @@
+		(os.FileMode(*fileStat.Mode)&os.ModeSetuid) != 0 {
+		return ErrSuidModeForbidden
+	}
+
+	return nil
+}
+
+func (p *Permit) DownloadFile(fileStat model.FileInfo) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	filePath := *fileStat.Path
+
+	if p.limits.FileTransfer.RegularFilesOnly && !utils.IsRegularFile(filePath) {
+		return ErrOnlyRegularFilesAllowed
+	}
+
+	if !utils.IsInChroot(filePath, p.limits.FileTransfer.Chroot) {
+		return ErrChrootViolation
+	}
+
+	if !utils.FileOwnerMatches(filePath, p.limits.FileTransfer.OwnerGet) {
+		return ErrFileOwnerMismatch
+	}
+
+	if !utils.FileGroupMatches(filePath, p.limits.FileTransfer.GroupGet) {
+		return ErrFileGroupMismatch
+	}
+
+	if !p.limits.FileTransfer.FollowSymLinks {
+		absolutePath, err := filepath.EvalSymlinks(filePath)
+		if err != nil {
+			return err
+		} else {
+			if absolutePath != filePath {
+				return ErrFollowLinksForbidden
+			}
+		}
+	}
+
+	if p.limits.FileTransfer.MaxFileSize > 0 {
+		fileSize := utils.FileSize(filePath)
+		if fileSize > 0 && p.limits.FileTransfer.MaxFileSize < uint64(fileSize) {
+			return ErrFileTooBig
+		}
+	}
+
+	return nil
+}
+
+func (p *Permit) BytesSent(n uint64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	countersMutex.Lock()
+	defer countersMutex.Unlock()
+
+	belowLimit = true
+	if n != 0 {
+		if deviceCountersLastH.bytesTransferred < math.MaxUint64-n {
+			deviceCountersLastH.bytesTransferred += n
+		}
+	}
+	if p.limits.FileTransfer.Counters.MaxBytesTxPerHour > 0 &&
+		deviceCountersLastH.bytesTransferred >= p.limits.FileTransfer.Counters.MaxBytesTxPerHour {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesTransferred < math.MaxUint64-n {
+			p.counters.bytesTransferred += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BytesReceived(n uint64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	countersMutex.Lock()
+	defer countersMutex.Unlock()
+
+	belowLimit = true
+	if n != 0 {
+		if deviceCountersLastH.bytesReceived < math.MaxUint64-n {
+			deviceCountersLastH.bytesReceived += n
+		}
+	}
+	if p.limits.FileTransfer.Counters.MaxBytesRxPerHour > 0 &&
+		deviceCountersLastH.bytesReceived >= p.limits.FileTransfer.Counters.MaxBytesRxPerHour {
+		belowLimit = false
+	}
+
+	p.countersMutex.Lock()
+	defer p.countersMutex.Unlock()
+	if n != 0 {
+		if p.counters.bytesReceived < math.MaxUint64-n {
+			p.counters.bytesReceived += n
+		}
+	}
+	return belowLimit
+}
+
+func (p *Permit) BelowMaxAllowedFileSize(offset int64) (belowLimit bool) {
+	if !p.limits.Enabled {
+		return true
+	}
+
+	if offset < 0 {
+		return true
+	}
+	if p.limits.FileTransfer.MaxFileSize > 0 &&
+		uint64(offset) >= p.limits.FileTransfer.MaxFileSize {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (p *Permit) PreserveModes(path string, mode os.FileMode) error {
+	if !p.limits.Enabled {
+		return nil
+	}
+
+	if (mode & S_ISUID) != 0 {
```

Comment:
This will never be the case.

Reply:
if you send mode=4755 it will

Reply:
Unless you have the "PreserveModes" config set, only the permission modes counts: https://github.com/mendersoftware/mender-connect/blob/1399abbbb0478ecc91e197b5ae2dbcc719236f7d/session/filetransfer.go#L471